ablySpec/Utilities.swift

```diff
@@ -0,0 +1,137 @@
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedBar).to(beNil())
+                    expect(receivedFooOrBar).to(equal(123))
+                    expect(receivedAll).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with:456)
+                    
+                    expect(receivedFoo1).to(equal(123))
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedBar).to(equal(456))
+                    expect(receivedFooOrBar).to(equal(456))
+                    expect(receivedAll).to(equal(456))
+                    
+                    eventEmitter.emit("qux", with:789)
+                    
+                    expect(receivedFooOrBar).to(equal(456))
+                    expect(receivedAll).to(equal(789))
+                }
+                
+                it("should only call once listeners once for its event") {
+                    eventEmitter.emit("foo", with: 123)
+
+                    expect(receivedBarOnce).to(beNil())
+                    expect(receivedAllOnce).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with: 456)
+                    
+                    expect(receivedBarOnce).to(equal(456))
+                    expect(receivedAllOnce).to(equal(123))
+
+                    eventEmitter.emit("bar", with: 789)
+                    
+                    expect(receivedBarOnce).to(equal(456))
+                    expect(receivedAllOnce).to(equal(123))
+                }
+                
+                it("should stop receiving events when calling off") {
+                    eventEmitter.offAll(listenerFoo1!)
+                    eventEmitter.emit("foo", with: 123)
+                    
+                    expect(receivedFoo1).to(beNil())
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedFooOrBar).to(equal(123))
+                    expect(receivedAll).to(equal(123))
+                    
+                    eventEmitter.off("foo", listener: listenerFooOrBar!)
+                    eventEmitter.emit("foo", with: 456)
+                    
+                    expect(receivedFoo2).to(equal(456))
+                    expect(receivedFooOrBar).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with: 789)
+                    
+                    expect(receivedFooOrBar).to(equal(789))
+                    
+                    eventEmitter.off("foo", listener: listenerAll!)
+                    eventEmitter.emit("foo", with: 111)
+                    
+                    expect(receivedAll).to(equal(789))
+                    
+                    eventEmitter.emit("bar", with: 222)
+                    
+                    expect(receivedAll).to(equal(222))
+                }
+
+                it("should receive only once after calling onAll and then once") {
```

Comment:
> Generally I would prefer that when you call off(), or removeListener you specify criteria - listener and/or event - and all registered handlers matching the given criteria are removed.

@paddybyers Ok, I agree and what is implemented in Ruby. I will add a quick test to ably-js to make sure that functionality is followed there, and add an issue to ably-java to ensure it matches that spec.  I will also quickly update the 0.8 spec to reflect this.